### PR TITLE
Disable program normalization (!= ==> !(==) rewriting builtin hack)

### LINKED ISSUE #851: "!=" comparator not matching Golang sample
I expected `$X != nil` to match `err != nil` in Golang, but get no match: https://semgrep.live/qNRz

Notably, `$X == nil` does match `err != nil`: https://semgrep.live/lBNy

I've been confused what equivalences are applied in the `==` case.

```
  "$X == nil" matches "err != nil" # unexpected
  "$X == nil"  matches "!(err == nil)" # unexpected? Could make an argument for this one
  "$X == nil" matches "!(err != nil)" # expected
  "$X == nil" matches "err == nil)" # expected
```

### PR DESCRIPTION
**Pull request by aryx**
This closes issue #851

This was a hack to demo some semantic matching but it actually works
in only one direction ... User-defined equivalences are better for that
and are more explicit.

Test plan:
test file included

### CODE REVIEW

**Comment by aryx** [comment]
This may lead to some test regressions in semgrep-rules where people may have relied on this builtin, but now they should really use explicit equivalences in the yaml file for that.

File: semgrep-core/matching/Semgrep_generic.ml [hunk 1]
```diff
@@ -222,15 +222,22 @@ let apply_equivalences equivs any =
 (*****************************************************************************)
 
 (*s: function [[Semgrep_generic.check2]] *)
-let check2 ~hook rules equivs file lang ast =
+let check2 ~hook rules equivs file _lang ast =
```
Comment (nbrahms) [inline semgrep-core/matching/Semgrep_generic.ml:225]: This was only here to support `Normalize_AST.normalize`, so we could undo the argument. 🤷

File: semgrep-core/tests/Test.ml [hunk 2]
```diff
@@ -94,7 +94,13 @@ let regression_tests_for_lang files lang =
     let rule = { R.
       id = "unit testing"; pattern; message = ""; severity = R.Error; 
       languages = [lang] } in
-    let equiv = [] in
+    let equiv = 
+     (* Python == is not the same than !(==) *)
```
Comment (nbrahms) [inline semgrep-core/tests/Test.ml:98]:
```suggestion
     (* Python != is not the same as !(==) *)
```

**Review by nbrahms — APPROVED** [review]

File: semgrep-core/tests/Test.ml [hunk 3]
```diff
@@ -94,7 +94,13 @@ let regression_tests_for_lang files lang =
...
+     if lang <> Lang.Python
```
Comment (nbrahms) [inline semgrep-core/tests/Test.ml:99]: FWIW, I think `basic_equivalences.yml` already excludes Python from the rewrite. Do we still need the language check here?

**Comment by brendongo** [comment]
Pulled in develop changes should make tests pass

**Comment by aryx** [comment]
How do you merge the changes in develop into this branch?
git merge develop  from develop_normalize branch? git rebase?

**Comment by brendongo** [comment]
I did a git merge. Normally I would rebase and force push but that would be messy for you to pull locally.

**Comment by brendongo** [comment]
Really just needed to pull these changes https://github.com/returntocorp/semgrep/pull/854 since pipenv released a new version and broke our CI

File: semgrep-core/matching/Semgrep_generic.ml (hunk 1 above)
Comment (aryx) [reply semgrep-core/matching/Semgrep_generic.ml:225]: Yep, except I've found a bug elsewhere that will require again this argument :)

File: semgrep-core/tests/Test.ml (hunk 3 above)
Comment (aryx) [reply semgrep-core/tests/Test.ml:99]: Oh, yeah, you're right! Actually I just realized the equivalence code does not respect the languages[] field
in the equivalence file ... I'll fix that in another diff.

**Pull request merged**